### PR TITLE
[HDF5] Fix Memory Errors

### PR DESCRIPTION
**Pull request by matekelemen**
## Description

Fix two memory errors related to matrix output.

1) an obvious segfault when attempting to query items within a container `rContainerItems.first`, when that container is empty.

2) an out of bounds read triggered by an uninitialized offset `local_dims[1]` within an array of offsets.

### CODE REVIEW

File: applications/HDF5Application/custom_io/hdf5_container_component_io.cpp [hunk 1]
```diff
@@ -699,18 +715,20 @@ void SetDataBuffer(Variable<Vector<double>> const& rVariable,
...
-        KRATOS_ERROR_IF(r_vec.size() != rData.size2())
-            << "Invalid dimension.\n";
-        for (std::size_t j = 0; j < rData.size2(); ++j)
-            rData(i, j) = r_vec(j);
-    }
+    if (!rContainerItems.empty()) {
+        rData.resize(rContainerItems.size(),
+                    rContainerItems.front()->GetValue(rVariable).size(), false);
+        #pragma omp parallel for
+        for (int i = 0; i < static_cast<int>(rContainerItems.size()); ++i) {
```
Comment (rubenzorrilla) [inline applications/HDF5Application/custom_io/hdf5_container_component_io.cpp]: Maybe we can take the chance to use our parallel utils in here.

File: applications/HDF5Application/custom_io/hdf5_container_component_io.cpp [hunk 2]
```diff
@@ -722,18 +740,20 @@ void SetDataBuffer(Variable<Matrix<double>> const& rVariable,
...
-        KRATOS_ERROR_IF(r_data.size() != rData.size2())
-            << "Invalid dimension.\n";
-        for (std::size_t j = 0; j < rData.size2(); ++j)
-            rData(i, j) = r_data[j];
-    }
+    if (!rContainerItems.empty()) {
+        rData.resize(rContainerItems.size(),
+                    rContainerItems.front()->GetValue(rVariable).data().size(), false);
+        #pragma omp parallel for
+        for (int i = 0; i < static_cast<int>(rContainerItems.size()); ++i) {
```
Comment (rubenzorrilla) [inline applications/HDF5Application/custom_io/hdf5_container_component_io.cpp]: Same comment.

Comment (matekelemen) [reply applications/HDF5Application/custom_io/hdf5_container_component_io.cpp]: done

**Comment by matekelemen** [comment]
Fixed another matrix output issue related to writing 0 matrix sizes when a rank doesn't have data to write.

**Review by RiccardoRossi — COMMENTED** [review]
is anyone reviewing this? if not...i don't see anything wrong with this (admittedly my review was very superficial...but if it was a bug i would accept this ASAP)

File: applications/HDF5Application/custom_io/hdf5_container_component_io.cpp [hunk 3]
```diff
@@ -132,10 +145,26 @@ class ContainerItemComponentIO
...
+                std::array<int, 2> {0, 0}
+                :
+                std::array<int,2> {static_cast<int>(rContainerItems.front()->GetValue(rComponent).size1()),
+                                   static_cast<int>(rContainerItems.front()->GetValue(rComponent).size2())};
+
+            // The sizes must be maxed across ranks to avoid
+            // those without data writing 0 sizes.
+            /// @todo This needs a revamp with DataCommunicator @matekelemen
+            int is_distributed;
+            MPI_Initialized(&is_distributed);
```
Comment (philbucher) [inline applications/HDF5Application/custom_io/hdf5_container_component_io.cpp]: I sm pretty sure this needs to be protected 

Have you compiled it without MPI?

**Comment by matekelemen** [comment]
This PR is superseded by a complete refactor of the application, which is under review.

**Pull request closed without merging**